src/charm.py

```diff
@@ -0,0 +1,595 @@
+            # Safe guard against transient unexpected error.
+            except Exception as err:  # pylint: disable=broad-exception-caught
+                logger.exception("Failed to update runner binary")
+                # Failure to download runner binary is a transient error.
+                # The charm automatically update runner binary on a schedule.
+                self.unit.status = MaintenanceStatus(f"Failed to update runner binary: {err}")
+                return
+            self.unit.status = MaintenanceStatus("Starting runners")
+            try:
+                self._reconcile_runners(runner_manager)
+                self.unit.status = ActiveStatus()
+            except RunnerError as err:
+                logger.exception("Failed to start runners")
+                self.unit.status = MaintenanceStatus(f"Failed to start runners: {err}")
+        else:
+            self.unit.status = BlockedStatus("Missing token or org/repo path config")
+
+    @catch_unexpected_charm_errors
+    def _on_upgrade_charm(self, _event: UpgradeCharmEvent) -> None:
+        """Handle the update of charm.
+
+        Args:
+            event: Event of charm upgrade.
+        """
+        logger.info("Reinstalling dependencies...")
+        self._install_deps()
+        self._start_services()
+
+        logger.info("Flushing the runners...")
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            return
+
+        runner_manager.flush()
+        self._reconcile_runners(runner_manager)
+
+    @catch_unexpected_charm_errors
+    def _on_config_changed(self, _event: ConfigChangedEvent) -> None:
+        """Handle the configuration change.
+
+        Args:
+            event: Event of configuration change.
+        """
+        try:
+            self._event_timer.ensure_event_timer(
+                "update-runner-bin", self.config["update-interval"]
+            )
+            self._event_timer.ensure_event_timer(
+                "reconcile-runners", self.config["reconcile-interval"]
+            )
+        except TimerEnableError as ex:
+            logger.exception("Failed to start the event timer")
+            self.unit.status = BlockedStatus(
+                f"Failed to start timer for regular reconciliation and binary update checks: {ex}"
+            )
+
+        if self.config["path"] != self._stored.path:
+            prev_runner_manager = self._get_runner_manager(
+                path=str(self._stored.path)
+            )  # Casting for mypy checks.
+            if prev_runner_manager:
+                self.unit.status = MaintenanceStatus("Removing runners from old org/repo")
+                prev_runner_manager.flush()
+            self._stored.path = self.config["path"]
+
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
+            self.unit.status = ActiveStatus()
+        else:
+            self.unit.status = BlockedStatus("Missing token or org/repo path config")
```

Comment:
Same comment here about having a not case first and continuing the execution flow.
i.e.
```suggestion
if not runner_manager:
    self.unit.status = BlockedStatus(...)
    return
... # continue execution flow
```

Reply:
Minor comments